src/postgres/abstract/AbstractPostgresClient.ts

```diff
@@ -0,0 +1,82 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { PostgreSQLManagementClient } from "@azure/arm-postgresql";
+import { PostgreSQLFlexibleManagementClient } from "@azure/arm-postgresql-flexible";
+import * as msRest from "@azure/ms-rest-js";
+import { IMinimumServiceClientOptions } from "vscode-azureextensionui";
+import { asAbstractDatabase, asFlexibleParameters, asSingleParameters, flexibleAsAbstractServer, singleAsAbstractServer } from "./maps";
+import * as Models from "./models";
+
+export class AbstractPostgresClient {
+    private postgresFlexibleClient: PostgreSQLFlexibleManagementClient;
+    private postgresSingleClient: PostgreSQLManagementClient;
+
+    constructor(credentials: msRest.ServiceClientCredentials, subscriptionId: string, options?: IMinimumServiceClientOptions) {
+        this.postgresFlexibleClient = new PostgreSQLFlexibleManagementClient(credentials, subscriptionId, options);
+        this.postgresSingleClient = new PostgreSQLManagementClient(credentials, subscriptionId, options);
+    }
+
+    async checkNameAvailability(serverType: Models.PostgresServerType, name: string) : Promise<Models.AbstractNameAvailability> {
+        switch (serverType){
+            case Models.PostgresServerType.Flexible:
+                return this.postgresFlexibleClient.checkNameAvailability.execute({name: name, type: "Microsoft.DBforPostgreSQL"});
+            case Models.PostgresServerType.Single:
+                return this.postgresSingleClient.checkNameAvailability.execute({name: name, type: "Microsoft.DBforPostgreSQL"});
+            default:
+                throw new Error("Service not implemented.");
+        }
+    }
+
+    async createOrUpdateFirewallRule(serverType: Models.PostgresServerType, resourceGroup: string, name: string, ruleName: string, rule: Models.AbstractFirewallRule): Promise<Models.AbstractFirewallRule> {
```

Comment:
The vast majority of these methods are just copy/pasted for both clients. Rather than duplicating the code, how about we do something like this:

```typescript
export type AbstractPostgresClient = PostgreSQLFlexibleManagementClient | PostgreSQLManagementClient;

export function createAbstractPostgresClient(serverType: PostgresServerType, clientInfo: ISubscriptionContext): AbstractPostgresClient {
    switch (serverType) {
        case PostgresServerType.Flexible:
            return createAzureClient(clientInfo, PostgreSQLFlexibleManagementClient);
        case PostgresServerType.Single:
            return createAzureClient(clientInfo, PostgreSQLManagementClient);
        default:
            throw new Error("Service not implemented.");
    }
}
```

As an example, `configurePostgresFirewall.ts` would then look like this:
```typescript
const client = createAbstractPostgresClient(treeItem.serverType, treeItem.root);
await client.firewallRules.createOrUpdate(resourceGroup, serverName, firewallRuleName, {
    startIpAddress: ip,
    endIpAddress: ip
});
```